### PR TITLE
docs/userguide/ext_modules.rst: New

### PR DESCRIPTION
**Pull request by mkoeppe**
## Summary of changes

Add documentation on `build_ext` as requested by @jaraco in https://github.com/python/cpython/issues/90539#issuecomment-1153274032

### Pull Request Checklist
- [ ] Changes have tests
- [ ] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_

[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.pypa.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

**Review by abravalheri — COMMENTED** [review]
Thank you very much @mkoeppe!

This is indeed a part of `setuptools` that is under documented and that sometimes requires users to read the `distutils` docs, which may be outdated and confusing.

I plan to add some example and a few remarks to the PR... I hope you don't mind.

**Comment by abravalheri** [comment]
I just added some changes in the spirit of improving the docs:

- I tried to provide a very simple example, just to show to the beginners how they can start adding extensions, and a few remarks.
- I also did some editorial/cosmetic changes (and tried to make sphinx more satisfied in terms of warnings).

I hope you don't mind the updates...

Please let me know if you would like to change anything, otherwise I plan to merge the PR very soon.

**Comment by abravalheri** [comment]
@Mergifyio rebase

**Comment by mkoeppe** [comment]
> I added some changes in the spirit of improving the docs

Thanks very much, looking great.

File: docs/userguide/ext_modules.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+Compiler and linker options
+===========================
+
+The command ``build_ext`` builds C/C++ extension modules.  It creates
+a command line for running the compiler and linker by combining
+compiler and linker options from various sources, as specified by
+`test_customize_compiler
+<https://github.com/pypa/setuptools/blob/main/setuptools/_distutils/tests/test_sysconfig.py>`_:
```
[the remark this answers is not in the document]
Comment (mkoeppe) [reply docs/userguide/ext_modules.rst]: That's fine. I might send a follow-up PR that adds some details (to support https://github.com/python/cpython/pull/92754) later.

[the remark this answers is not in the document]
Comment (abravalheri) [reply docs/userguide/ext_modules.rst]: Thank you very much @mkoeppe.

Now that this document exist, I also plan to move the [section about Cython](https://setuptools.pypa.io/en/latest/userguide/distribution.html#distributing-extensions-compiled-with-cython) there.

**Pull request merged**